### PR TITLE
PF-393 Personal Environments Dev Guide

### PR DESCRIPTION
**Pull request by jaycarlton**
An attempt at an end-to-end guide for setting up and administering a new personal environment, aimed at new developers with no assumed context on any of the technologies.

So far, I've got everything I know, and a bunch of things I don't, so please don't take anything odd at face value.

Calling this reviewable as a first useful increment. I'll add TODOs and questions I have in comments.

### CODE REVIEW

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (tnadler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Nit (maybe incorrect) - is a personal environment contained by a namespace or is it just a namespace?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (tnadler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Not sure where this belongs, but early on @wchamber explained to me the difference in how we use terraform and helm which was really useful. Basically, terraform brings up and configures all resources surrounding the core service (storage, iam, etc), while helm basically just provides environment variables to the running service. Also worth early on stating that we have this hierarchy in both cases where there is a generic thing in one repo and then a specific layer you need to add on in another, private repo.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (tnadler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I would not paste this in here - I am about to change this and I am sure it will continue to evolve. Perhaps better to point at an existing personal file?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (tnadler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: You need to get review first, and best way to do that is to post in dsp-devops-champion slack channel.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (tnadler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'd add something about how to get started in swagger ui. In particular, it confused me that for resource buffer I needed to glogin as a service account, while for wsm it had to be my broad account (that may be a bug somewhere)

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I don't really know. I think of the namespace as the boundary of the environment, like how the EU has borders but isn't the borders. But I'm just guessing.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Whoa, so helm is almost like Zookeeper in that model. I thought it also deployed the services pods via their helm charts.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Either way someone needs to update this when it changes. I prefer an end-to-end, self-contained-ish guide, but I'm certainly linking to plenty of other things.

I can indicate that the `tfvars` should line up with the variables specified in the modules, assuming those are easy to get to from here.

File: .gitignore
Comment (gjuggler) [inline .gitignore]: Is this gitignore copied from somewhere, or auto-generated? May be worth adding a comment line pointing back to the origin for provenance if so.

File: README.md
Comment (gjuggler) [inline README.md]: May be simpler, maintenance-wise, to just link to the dev-guides folder. I don't think there's too much value in individually indexing from this page — especially since in the long run, this repo will likely house lots more information about governance, processes, etc.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: File naming nit: prefer lowercase for a less-shouty directory listing, personal-environments.md

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'm not sure this is helpful, or in the long run, accurate. There's nothing stopping from the WM service running on multiple pods in the future.

I think a more important aspect to highlight is that a given personal environment can have any (sub)set of Terra services enabled or disabled, depending on which services a developer intends to test most frequently.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: style nit: this is shared documentation and should never use first person. Remove parenthetical.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: nit: I'd probably bump the book recommendations to a separate top-level section at the end, "Additional Reading". For introductory links with immediately-accessible material, we're better off linking out to the DevOps team's getting started guides: https://docs.dsp-devops.broadinstitute.org/best-practices-guides/terraform and https://docs.dsp-devops.broadinstitute.org/kubernetes/kubectl-getting-started

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Nit: I won't add further comments but please scrub through the rest of the file and remove first-person language.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'm not sure this belongs. This duplicates information in the Verily onboarding guide (and maybe / likely some Broad-specific materials as well), which means it's another place we have to actively keep in sync any changes.

I'd rather hold off from documenting this elsewhere, and if we do feel it's important to include in this public repo, I'd eventually create a separate dev-onboarding.md guide.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Helm does more than providing environment variables to the running service - it's our template for all configuration/specification within k8s. The line between terraform & helm is what each is more targeted to do: terraform configure cloud resources & helm for everything in k8s land.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: nit: `Deployments` is pretty overloaded. Are we talking about what makes up a "k8s deployment" https://kubernetes.io/docs/concepts/workloads/controllers/deployment/, or a logical Terra deployment, or a logical deployment of a single Terra service?
Maybe something like, "each Terra service is described in their corresponding helm chart by a series of related k8s resources, the primary one being the k8s deployment... etc."

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: > will depend on (but not deploy)

I think this isn't true – the `terra_apps` specifies the applications which will be enabled in the given environment. A `true` value here causes Terraform to create the necessary cloud infrastructure for the application.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'm not going to be a great resource on exactly what our helm does. Maybe appeal to authority and link to some helm site documentation :)

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: This may be useful context for a dev to know what to expect when getting a PR review in the Terraform repo, but it's not unique to personal environments. It may be better to link to the canonical source for expected PR flow maintained by the DevOps team, over at https://github.com/broadinstitute/terraform-ap-deployments#workflow

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: s/our deployment/a personal environment/

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: somewhere, it's worth linking to https://github.com/broadinstitute/terraform-ap-deployments#workflow which is canon for this process. Fine to restate parts of it here, but good to make folks aware of that documentation.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'm not sure this section is really relevant to personal environments. It's mildly interesting, but probably not worth keeping, in the interest of having focused & concise documentation. IMO, these dev guides should be kept as short as possible in order to (1) reduce maintenance burden and (2) increase the likelihood that they're read by developers.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: For at least the example, maybe stick to just workspacemanager - otherwise it might confuse folks about whether their WM is talking to their janitor or buffer (it isn't by default)
I'm not sure you need crljanitor or buffer. Are you working on either?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: What happens if DevOps changes the file structure in the future? This may be a bit too much detail to have to keep in sync with the code.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: This section is almost identical in name to the prior section. Is there a clearer distinction between the two?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I'm a little surprised this is necessary, but maybe DevOps has been doing it for me this whole time.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Prior headings at this level were numbered "Step 1" and "Step 2". Should this be "Step 3: Creating the namespace"?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (wchamber) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: fwiw, they're configured differently somewhat on purpose. buffer is never supposed to be user visible, so user accounts should not have access to its endpoints. WM should be accessible by users, but we don't want our dev instances open to the public. (Last I heard this was going to be blocked by VPN, but I guess that's changed.)

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: s/environments/applications/

And note, the Helmfile does not have any bearing on resources getting generated by Terraform. I think we need to manually to keep the Terraform app-enablement in sync with the Helmfile app-enablement.

**Review by wchamber — APPROVED** [review]
lgtm, but probably the more eyes the better :)

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [inline docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I feel like the document structure begins to break down here. What is the final step to creating a personal environment? My experience was that once I asked DevOps to use ArgoCD to "create" my environment application, the services were deployed and I could see the namespace and services running within kubernetes. I was also able to interact with ArgoCD itself after that point, including making a manual re-sync of my personal environment.

Maybe a slight re-structure would give some clarity, something like:
```
## Step 4: ArgoCD sync
## Appendices
### Using `kubectl` (though IMO we should probably just defer to DevOps team's docs here, unless there's something missing from there.
```

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: +1, I didn't have permissions to do this myself when I created mine, I had to ask DevOps to create it for me.

**Review by gjuggler — APPROVED** [review]
Thanks Jay for quickly drafting this guide! Left a handful of nits & suggestions. The one I probably feel most strongly about is the use of personal pronouns; use of "I", "we", "our", etc. should be avoided where possible for documentation like this.

Approving now, but please address comments as you see fit before merging.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Yes, I just rediscovered it, and it's pretty comprehensive and easy to follow.

File: .gitignore
Comment (jaycarlton) [reply .gitignore]: It's from workspace manager, which is probably not the right place to steal it. I just need .idea for right now.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I've misunderstood what these entries do I think. More cargo cult programming.

File: README.md
Comment (jaycarlton) [reply README.md]: A high-level descriptive index might make sense at some point, but I can link to the directory for now.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: OK, though I thought that was the convention for things like README, LICENSE, etc.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: If that's the standard, OK. I just find the alternative to require passive voice and awkward constructions, and not to be any clearer.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: we are now linking there. As with other places, there's some repetition here to make it a self-contained-ish end-to-end reference.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I wanted to be able to say Here's how you build your images, and here's where they're put into pods, then releases, then on down the line. I didn't see how to≠ to talk about the helmfile without at least pointing to helm charts. This his a spot that's really hard to get context on, because you never see the basic k8s files in our workflow. So I can't talk about images or pods really without getting into the helm chart.

Maybe the scope is broader than personal environments, because I wanted to make it more accessible for new people. Maybe another guide to how all the yaml gets merged and interpolated and how to trace it through the system would help?

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I don't know how much choice they have in the structure. I'm definitely pointing to the only things I know here, and it's not very deep at this point. 

The exact structure isn't terribly important to the narrative.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: renamed

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Removed references to buffer and crljanitor.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: good catch

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: good point

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I've done this twice now. The trick is to use a mechanical keyboard.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Do you still have access to this ArgoCD application? I get a 403 now when I try to go to https://ap-argocd.dsp-devops.broadinstitute.org/applications/terra-app-generator

Yes, the document falls apart because this was as far as I'd gotten.

I'm thinking I could add screenshots for looking at the deployment and related things in the Cloud Console. The GKE interface seems fairly stable at this point, but both that and `kubectl` are well documented externally.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (jaycarlton) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: I propose we do a separate dev guide for working with Swagger UI. It has more to do with services than with environments from what I gather.

File: docs/dev-guides/PERSONAL-ENVIRONMENTS.md
Comment (gjuggler) [reply docs/dev-guides/PERSONAL-ENVIRONMENTS.md]: Access to ArgoCD is protected behind the Broad firewall – I can see that page if I connect to the non-split VPN.

It may be useful to include some helpful ArgoCD and/or Cloud Console starting-off links. But my instinct would be to merge the guide with this scope sooner than later. We can either add to it over time if it proves useful-but-incomplete, or split out more of a getting-started guide that focuses more on the fundamentals of working w/ the current Terraform/Atlantis/Helm/ArgoCD stack.

**Pull request merged**
